src/io/flutter/vmService/VMServiceManager.java

```diff
@@ -630,7 +631,16 @@ public void received(JsonObject object) {
             ret.complete(null);
           }
           else {
-            ret.complete(object.get(fpsField).getAsDouble());
+            final double fps = object.get(fpsField).getAsDouble();
+            // Defend against invalid refresh rate for Flutter Desktop devices (0.0).
+            if (fps == 0.0) {
```

Comment:
Instead of putting this check in the `VMServiceManager.java` class - which currently doesn't know about the UI - perhaps move this check to the UI portion of the app which consumes this API? We'd want to make sure that we didn't perform the same check in multiple places.

Reply:
not to bike shed but anything under 24 fps could be considered invalid as 24fps is the lowest frame rate anyone considers smooth.

Reply:
Ah. I originally placed it in the VMServiceManager because I was afraid of it being called multiple times by the UI code that updates frequently. I now realize that even where it is in the VMServiceManager, it is called multiple times because it is triggered on each hot restart. Looks like we need to introduce an additional layer that manages this notification and prevents it from being shown multiple times.

Reply:
Add two classes to resolve this: DisplayRefreshRateManager and NotificationManager. Ptal.

Reply:
used 24 fps as the threshold